### PR TITLE
[mlir][loops] Reland Refactor LoopFuseSiblingOp and support parallel fusion #94391

### PR DESCRIPTION
**Pull request by srcarroll**
The refactor had a bug where the fused loop was inserted in an incorrect location.  This patch fixes the bug and relands the original PR https://github.com/llvm/llvm-project/pull/94391.

This patch refactors code related to LoopFuseSiblingOp transform in attempt to reduce duplicate common code. The aim is to refactor as much as possible to a functions on LoopLikeOpInterfaces, but this is still a work in progress. A full refactor will require more additions to the LoopLikeOpInterface.

In addition, scf.parallel fusion support has been added.

### CODE REVIEW

**Comment by srcarroll** [comment]
for those who already reviewed the original PR, the only changes i made to fix the bug are in this commit https://github.com/llvm/llvm-project/pull/97607/commits/4e4a96e376aaf778d013d384dc3c2b9dab405f35

**Comment by srcarroll** [comment]
@qedawkins please confirm the test shows what you expect

**Review by qedawkins — APPROVED** [review]
LGTM, thanks for the quick fix!

File: mlir/lib/Interfaces/LoopLikeInterface.cpp [hunk 1]
```diff
@@ -113,3 +115,58 @@ LogicalResult detail::verifyLoopLikeOpInterface(Operation *op) {
...
+  auto sourceRegion = source.getLoopRegions().front();
+
+  FailureOr<LoopLikeOpInterface> maybeFusedLoop =
+      target.replaceWithAdditionalYields(rewriter, source.getInits(),
+                                         /*replaceInitOperandUsesInLoop=*/false,
+                                         newYieldValuesFn);
+  if (failed(maybeFusedLoop))
+    llvm_unreachable("failed to replace loop");
+  LoopLikeOpInterface fusedLoop = *maybeFusedLoop;
+  rewriter.moveOpBefore(fusedLoop, source);
```
Comment (Hardcode84) [inline mlir/lib/Interfaces/LoopLikeInterface.cpp:141]: Can you add code comment, why this is needed?

**Review by Hardcode84 — APPROVED** [review]

File: mlir/lib/Interfaces/LoopLikeInterface.cpp (hunk 1 above)
Comment (srcarroll) [reply mlir/lib/Interfaces/LoopLikeInterface.cpp:143]: done

**Pull request merged**